feat(`server`): agnostic http request handler

fallenfreq opened the pull request:
Closes #4280 / 
Closes T-26

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

1. You can now add custom servers without them having to have the same req.query type as express.
ParsedQs interface has been removed for this because its only use becomes redundant if you can't guarantee that it can be passed to URLSearchParams

2. The middleware method that you pass to an adapter will now pass the correct types through to the "req" and "res" arguments.

3. Added the necessary tests to test the new features.

## ✅ Checklist

- ✅  I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- ✅  If necessary, I have added documentation related to the changes made.
- ✅ I have added or updated the tests related to the changes made.

KATT reviewed (changes requested):
This unfortunately fails all tests

KATT reviewed (dismissed):
This unfortunately fails all tests

fallenfreq commented:
> This unfortunately fails all tests

Sorry about that. It should be passing all tests and checks now apart from the 2 checks I'm not Authorized to run.

KATT commented:
hey @Sheraff, could you review this? :)

Sheraff commented on packages/server/src/adapters/node-http/nodeHTTPRequestHandler.ts:
```diff
@@ -27,9 +27,17 @@ export async function nodeHTTPRequestHandler<
-    const query = opts.req.query
-      ? new URLSearchParams(opts.req.query as any)
-      : new URLSearchParams(opts.req.url!.split('?')[1]);
+    // a framework could pre-parse query in a way that could crash URLSearchParams
+    const paramsFromQuery = () =>
+      new URLSearchParams(opts.req.url?.split('?')[1]);
+    let query: URLSearchParams;
+    try {
+      query = opts.req.query
+        ? new URLSearchParams(opts.req.query as any)
+        : paramsFromQuery();
+    } catch (err) {
+      query = paramsFromQuery();
+    }
```
I think I do understand the need for this PR, but this feels a bit sub-optimal since there could be cases where *every* procedure ends up throwing (from `new URLSearchParams(opts.req.query as any)`) before being recovered by the `catch`.

It might be a little nitpicky but aren't there alternative implementations that wouldn't be require this? I don't want to speak for KATT here but maybe a boolean option on the server configuration instead of a `try/catch` on every procedure call?

juliusmarminge replied:
@fallenfreq friendly bump

fallenfreq replied:
@Sheraff @juliusmarminge Hey, I've been camping with no signal for a few weeks and now I'm on a work trip for the next two weeks. I'll be able to look at this again when I'm home towards the end of the week starting July the 10th.

I didn't consider an extra parameters originally because in my case, it would only fail on a rare occurrences but it makes sense considering there are cases where it could fail on every request.

Would you have the parameter and keep the try catch block for cases where it has not been set? Or perhaps catch and re throw an error letting the user know they should be setting the parameter?

How do the names isQueryInvalid or invalidTrpcQuery sound?

juliusmarminge commented on pnpm-lock.yaml:
```diff
@@ -13762,14 +13809,14 @@ packages:
         optional: true
     dependencies:
-      '@typescript-eslint/parser': 6.2.1(eslint@8.40.0)(typescript@5.1.3)
+      '@typescript-eslint/parser': 5.59.2(eslint@8.40.0)(typescript@5.1.3)
```
why is this getting downgraded lol

fallenfreq replied:
There are other changes in the file that where needed for the new example "examples/minimal-custom-server" to pass all tests. I can only assume the file has been updated several times since I made the amends

KATT reviewed (commented):
Hey, sorry for the slow response on this!

Could you please?:

- remove the example project - we have enough of them haha
- add a test, probably in https://github.com/trpc/trpc/tree/main/packages/tests/server/adapters

see https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md#testing

fallenfreq commented:
> Hey, sorry for the slow response on this!
> 
> Could you please?:
> 
> * remove the example project - we have enough of them haha
> * add a test, probably in https://github.com/trpc/trpc/tree/main/packages/tests/server/adapters
> 
> see https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md#testing

Hey, I have removed the example and refactored it into a test

kamilogorek commented on packages/server/src/adapters/node-http/nodeHTTPRequestHandler.ts:
```diff
@@ -28,9 +28,12 @@ export async function nodeHTTPRequestHandler<
+    // a framework could pre-parse query in a way that could crash URLSearchParams
+    // invalidTrpcQuery can be added to a request object to let trpc know
+    const query: URLSearchParams =
+      opts.req.query && !opts.req.invalidTrpcQuery
```
I'd name it by the action it'll perform, not by the "state" the request is in. Eg. `forceQueryFromUrl` or `forceUrlQuery` provides some meaning in itself, while `invalidTrpcQuery` doesn't tell you what will happen, but only indicates that something is "off".

fallenfreq replied:
@kamilogorek I went with invalidTrpcQuery to differentiate the query that trpc uses internally from the query used by the framework. I'm happy to go with something like forceTrpcQueryFromUrl. My original thinking was that it leaks implementation details to the user that could change. For example, not all frameworks have a req.url property to make a query from although the ones I can think of already have a req.query parameter that is compatible with the trpc query and if they didn't, then they could just add a req.url property to generate the query from.

KATT commented:
cc @fallenfreq could you get this to point to `next` instead? we're not really doing new features on `main` anymore

sorry for the delay on this!

fallenfreq commented:
> cc @fallenfreq could you get this to point to `next` instead? we're not really doing new features on `main` anymore
> 
> sorry for the delay on this!

Yeah, I'll be working on a project that uses trpc again very soon so i'll do it then.

fallenfreq commented:
Hey @KATT, This is ready for review again. The original disputed property name is no longer required with trpc v11. I've updated the changes and tests to work with v11 so hopefully this is good to be merged into next now.

KATT commented on packages/tests/server/adapters/custom.test.ts:
Please just delete this, we have plenty of tests that test this path

fallenfreq replied:
Will do, this was originally testing the now redundant forceTrpcQueryFromUrl property as well but it's still testing that query parameters that don't match the express type don't throw and that the handler middleware has the types passed correctly. The other tests are just testing the implementation of the test itself.

fallenfreq replied:
Hi @KATT, This was done 3 weeks ago, I know it said there were two extra failing tasks but I think that was just due to a token generation limit being reached for some of the tests. I've just updated it and those failing tests have passed.
